### ISSUE TITLE
Is there a way to acquire "lost" unreleased locks by other clients?

### ISSUE THREAD

**Issue by theist** [cirello-io/dynamolock#44]
Maybe I'm just doing something wrong. I've created a small program based on the example provided that acquire a lock, but it does not close the client because I want to control a program that runs in two seconds every 5 minutes or so but needs to lock the execution in other nodes during more than 5 minutes.

My plan was to create a lock with a lease of 6 minutes and check for the lock before start the execution of the rest of the code.

But in my tests, unreleased locks are left there and never expire:

first execution:
```
[main] 14:35:38 Starting client
[main] 14:35:38 Acquiring lock
[dyna] 14:35:38 Call GetItem to see if the lock for  lock  = test  exists in the table
[dyna] 14:35:38 starting heartbeats
[dyna] 14:35:39 Lock   <nil>
[dyna] 14:35:39 Captured new Lock
[dyna] 14:35:39 Acquiring a new lock or an existing yet released lock on  lock = test
[main] 14:35:39 lock content: 
[main] 14:35:39 lock owner: test owner
[main] 14:35:39 expired?  false
[main] 14:35:39 done
```

next execution, the lock is in the DynamoDB table but it is not released and the lease is for 1m

```
./main 
[main] 14:35:46 Starting client
[main] 14:35:46 Acquiring lock
[dyna] 14:35:46 Call GetItem to see if the lock for  lock  = test  exists in the table
[dyna] 14:35:46 starting heartbeats
[dyna] 14:35:46 Lock   &{{0 0} 0xc000196d00 test [] test owner false false <nil> {13786376802911714670 247280565 0xcea6a0} ZGNHp7OU5dRRDsHGz05NbxQ6mNq2Rab4 60000000000 map[]}
[dyna] 14:35:46 Existing Lock Owner:  test owner
[dyna] 14:35:46 New Lock Owner:  test owner
[dyna] 14:35:46 Existing Lock isExpired:  false
[dyna] 14:35:46 Existing Lock lookupTime:  2019-09-01 14:35:46.758971758 +0200 CEST m=+0.247280565
[main] 14:35:46 Error acquiring lock: Didn't acquire lock because it is locked and request is configured not to retry.
```

This was expected but, next time I execute it the lock remains in the table, and unreselased. This is normal because there's no code runing to release it. The lease was for 1m but the execution happens 10 minutes after the creation of the lock: 
```
./main 
[main] 14:46:51 Starting client
[main] 14:46:51 Acquiring lock
[dyna] 14:46:51 starting heartbeats
[dyna] 14:46:51 Call GetItem to see if the lock for  lock  = test  exists in the table
[dyna] 14:46:52 Lock   &{{0 0} 0xc000198820 test [] test owner false false <nil> {13786377517353580211 328045090 0xcea6a0} ZGNHp7OU5dRRDsHGz05NbxQ6mNq2Rab4 60000000000 map[]}
[dyna] 14:46:52 Existing Lock Owner:  test owner
[dyna] 14:46:52 New Lock Owner:  test owner
[dyna] 14:46:52 Existing Lock isExpired:  false
[dyna] 14:46:52 Existing Lock lookupTime:  2019-09-01 14:46:52.088782515 +0200 CEST m=+0.328045090
[main] 14:46:52 Error acquiring lock: Didn't acquire lock because it is locked and request is configured not to retry.
```

Looks to me that the library would consider locked any existing lock that has not been properly released, regardless of the lease time. 

But perhaps I'm missing something. 

My code:
```golang
package main

import (
	"log"
	"os"
	"time"

	"cirello.io/dynamolock"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/dynamodb"
)

func main() {
	log.SetPrefix("[main] ")
	log.SetFlags(log.Ltime)
	svc := dynamodb.New(session.Must(session.NewSession(&aws.Config{
		Region: aws.String("eu-west-1"),
	})))

	log.Println("Starting client")
	c, err := dynamolock.New(svc,
		"Locks",
		dynamolock.WithLeaseDuration(1*time.Minute),
		dynamolock.WithLogger(log.New(os.Stdout, "[dyna] ", log.Ltime)),
		dynamolock.WithPartitionKeyName("lock"),
		dynamolock.WithOwnerName("test owner"),
	)
	if err != nil {
		log.Fatal(err)
	}

	log.Println("Acquiring lock")
	lockedItem, err := c.AcquireLock("test",
		dynamolock.ReplaceData(),
		dynamolock.FailIfLocked(),
	)
	if err != nil {
		log.Fatal("Error acquiring lock: ", err)
	}

	log.Println("lock content:", string(lockedItem.Data()))
	log.Println("lock owner:", lockedItem.OwnerName())
	log.Println("expired? ", lockedItem.IsExpired())

	log.Println("done")
}
```

**Comment by ucirello** [comment]
TL;DR: if you remove `dynamolock.FailIfLocked()`, the code will probably do what you want it to.

> I've created a small program based on the example provided that acquire a lock, but it does not close the client because I want to control a program that runs in two seconds every 5 minutes or so but needs to lock the execution in other nodes during more than 5 minutes.

It seems what you are describing here is the leader election pattern. Is this correct?

> But in my tests, unreleased locks are left there and never expire:
> This was expected but, next time I execute it the lock remains in the table, and unreselased. This is normal because there's no code runing to release it. The lease was for 1m but the execution happens 10 minutes after the creation of the lock:

The lease duration is stored in the lock entry. Even if you reconfigure the client to use a different time in a second run, the stored lease time in the first one is what the dynamolock client is going to use to test for the lock acquisition.

This lock does not rely on a timestamp to test for lock loss. Assume client A and B, both with the same heartbeat frequency and lock lease duration. When the client A grabs a lock and stop doing heartbeats, client B is going to try to grab the lock, fail and then wait at least for the duration of the lease before trying to grab the lock again.

You should `defer lock.Close()` if you want to force the lock to be released. The client does not have a way to know when your application has failed or not. You should either close the client or release the lock. 

If the entry in the dynamoDB is a problem, you should consider using [WithDeleteLockOnRelease](https://godoc.org/cirello.io/dynamolock#WithDeleteLockOnRelease) and then make sure the lock is released.

In any case, competing clients wait for the lease duration before trying to grab a lock. `dynamolock.FailIfLocked()` disable this feature. Please consider trying your code without it.

**Comment by theist** [comment]
Oh, I see. Leftover locks are watched during the lease time, regardless of when they're created

```
[main] 09:17:02 Starting client
[main] 09:17:02 Acquiring lock
[dyna] 09:17:02 Call GetItem to see if the lock for  lock  = test  exists in the table
[dyna] 09:17:02 starting heartbeats
[dyna] 09:17:02 Lock   &{{0 0} 0xc000184340 test [] test owner false false <nil> {13786449039733041696 275380792 0xce96a0} ZGNHp7OU5dRRDsHGz05NbxQ6mNq2Rab4 60000000000 map[]}
[dyna] 09:17:02 Existing Lock Owner:  test owner
[dyna] 09:17:02 New Lock Owner:  test owner
[dyna] 09:17:02 Existing Lock isExpired:  false
[dyna] 09:17:02 Existing Lock lookupTime:  2019-09-02 09:17:02.52534736 +0200 CEST m=+0.275380792
[dyna] 09:17:02 Sleeping for a refresh period of  1s
[dyna] 09:17:03 Call GetItem to see if the lock for  lock  = test  exists in the table
...
[dyna] 09:18:02 Sleeping for a refresh period of  1s
[dyna] 09:18:03 Call GetItem to see if the lock for  lock  = test  exists in the table
[dyna] 09:18:03 Lock   &{{0 0} 0xc000184340 test [] test owner false false <nil> {13786449104886106764 60930194610 0xce96a0} ZGNHp7OU5dRRDsHGz05NbxQ6mNq2Rab4 60000000000 map[]}
[dyna] 09:18:03 Existing Lock Owner:  test owner
[dyna] 09:18:03 New Lock Owner:  test owner
[dyna] 09:18:03 Existing Lock isExpired:  false
[dyna] 09:18:03 Existing Lock lookupTime:  2019-09-02 09:18:03.180161164 +0200 CEST m=+60.930194610
[dyna] 09:18:03 Acquiring an existing lock whose revisionVersionNumber did not change for  lock  partitionKeyName= test
[main] 09:18:03 lock content: 
[main] 09:18:03 lock owner: test owner
[main] 09:18:03 expired?  false
[main] 09:18:03 done
```

> It seems what you are describing here is the leader election pattern. Is this correct?

Not sure. I'm more Sysadm / DevOps guy And I'm afraid my needs are more simple. I'm just triying to create a distributed version of [`flock`](https://linux.die.net/man/1/flock) so, a program that `os.Exit(1)` if it fails to get a lock and plan to use it in cron like `dlock && <struff to lock>` to get sure only one node executes it.

So for this, I have several problems with this library. First I need to leave the lock unreleased because the locker program exits before the lock time, and also would need that the second iteration from the same node should be able to recover the lock w/o waiting.

For the second I had a plan to add an `AutomaticallyAquireOnSameOwner()` Aquire option but I thing the lease time stuff is hard to overcome with my limited Go Skills

Thanks for the kind explanation and you can consider this closed
